File: src/mixins/WorkDays.js

/*
 * Mixin to compute fitness regarding number
 * of work days in work period.
 */

'use strict';

const WorkDays = (Base, props) => class extends Base {
  fitness() {
    let curr1
      , result = super.fitness()
      , employeesIter = props.employees();

    // Iterate through employees
    while (curr1 = employeesIter.next()) {
      let curr2
        , workDays = 0
        , periodDaysIter = props.periodDays();

      // Iterate through period days
      while (curr2 = periodDaysIter.next()) {
        // Check for work day
        if (this[props.shiftIndex(curr1, curr2)]) {
          workDays += 1;
        }
      }

      // Only exact work days produces 0 fitness decrease
      result += -Math.abs(props.opts.workDays - workDays);
    }

    return result;
  };
}

WorkDays.requiredOpts = {
  workDays: true
};

module.exports = WorkDays;

File: src/core/main/java/workschedule/schedule/fitnesses/WorkDays.java
package workschedule.schedule.fitnesses;

import workschedule.Chromosome;
import workschedule.schedule.ScheduleDay;
import workschedule.schedule.ScheduleProperties;

import java.util.Iterator;

public final class WorkDays extends AbstractFitness {
    public WorkDays(final ScheduleProperties props) {
        super(props);
    }

    @Override
    public int fitness(final Chromosome chr) {
        int result = 0;

        Iterator<Integer> employeeIterator = props.employees();

        // Iterate through all employees
        while (employeeIterator.hasNext()) {
            int workDays = 0;

            Integer employee = employeeIterator.next();

            Iterator<ScheduleDay> scheduleDayIterator = props.scheduleDays();

            // Iterate through schedule days
            while (scheduleDayIterator.hasNext()) {
                if (chr.getParam(props.getShiftIndex(employee, scheduleDayIterator.next()))) {
                    workDays++;
                }
            }

            // Only exact number of work days produces 0 fitness decrease
            result -= Math.abs(opts.getWorkDays() - workDays);
        }

        return result;
    }
}
